skl2onnx/operator_converters/KMeans.py

```diff
@@ -0,0 +1,122 @@
+    Symbols:
+    
+    * *l*: # of examples.
+    * *n*: # of features per input example.
+    * *X*: input examples, l-by-n tensor.
+    * *C*: centroids, k-by-n tensor.
+    * $C^2$: 2-norm of all centriod vectors, its shape is ``[k]``.
+    * *Y*: 2-norm of difference between examples and centriods, *l-by-k* tensor.
+      The value at i-th row and k-th column row, ``Y[i,k]``,
+      is the distance from example *i* to centroid *k*.
+    * *L*: the id of the nearest centroid for each input example, its shape is ``[l]``.
+    
+    ::
+    
+        .------------------------------------------------------.
+        |                                                      |
+        |                                                      v
+        X [l, n] --> ReduceSumSquare --> X^2 [l]             Gemm (alpha=-2, transB=1) <-- C [k, n]
+                                          |                    |
+                                          |                    v
+                                          `------> Add <---- -2XC' [l, k]
+                                                    |
+                                                    v
+                                                    Z [l, k] ----------> Add <------------C^2 [k]
+                                                                          |
+                                                                          v
+                                                  L [l] <--- ArgMin <---  Y2 [l, k] --> Sqrt --> Y2 [l, k]
+    
+    *scikit-learn* code:
+    
+    ::
+    
+        X = data
+        Y = model.cluster_centers_
+        XX = row_norms(X, squared=True)
+        YY = row_norms(Y, squared=True)
+        distances = safe_sparse_dot(X, Y.T, dense_output=True)
+        distances *= -2
+        distances += XX[:, numpy.newaxis]
+        distances += YY[numpy.newaxis, :]
+        numpy.sqrt(distances, out=distances)
+    """
+    op = operator.raw_operator
+    variable = operator.inputs[0]
+    N = variable.type.shape[0]
+    
+    # centroids        
+    shapeC = list(op.cluster_centers_.shape)
+    nameC = scope.get_unique_variable_name('centroid')
+    container.add_initializer(nameC, onnx_proto.TensorProto.FLOAT,
+                              shapeC, op.cluster_centers_.flatten())    
+
+    nameX2 = scope.get_unique_variable_name('X2')
+    nameX = operator.inputs[0].full_name
+    container.add_node('ReduceSumSquare', [nameX], [nameX2], axes=[1], keepdims=1,
+                        name=scope.get_unique_operator_name('ReduceSumSquare'))
+    
+    # Compute -2XC'
+    zero_name = scope.get_unique_variable_name('zero')
+    zeros = np.zeros((N, ))
+    container.add_initializer(zero_name, onnx_proto.TensorProto.FLOAT,
+                              list(zeros.shape), zeros)   
+    
+    if container.target_opset < 5:
+        attrs['broadcast'] = 1
+        op_version = 1
```

Comment:
Given this is first time to support this op, probably we don't care about opset less than 7.

Reply:
I moved all operators which dépends on version in _apply_operator. kmeans and gemm operator should be two independant problem now.